### PR TITLE
[Traceable FSDP2] Use DTensor.from_local() in _from_local_no_grad when compile

### PR DESCRIPTION
**Pull request by yf225**
As discussed before, for now Dynamo is not able to support DTensor constructor, and instead we have to use `DTensor.from_local()`.

This won't affect eager and it's a compile-only change.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @d4l3k

### CODE REVIEW

**Review by awgu — APPROVED** [review]
Sounds good!

File: torch/distributed/_composable/fsdp/_fsdp_common.py [hunk 1]
```diff
@@ -117,20 +117,29 @@ def _from_local_no_grad(
...
-        # variable, e.g. with `view_as()`, since this is not differentiable
-        local_tensor,
-        device_mesh,
-        placements,
-        shape=global_size,
-        dtype=local_tensor.dtype,
-        requires_grad=local_tensor.requires_grad,
-        stride=global_stride,
-    )
+    if not torch._dynamo.compiled_autograd.compiled_autograd_enabled:
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_common.py:123]: Actually, I wonder if we can do `if not torch.is_grad_enabled():`?

I think `_from_local_no_grad` on the hot path is mainly for converting sharded gradient to `DTensor` in `foreach_reduce`, which is outside autograd (since `@torch.no_grad()` decorator).

Comment (yf225) [reply torch/distributed/_composable/fsdp/_fsdp_common.py:123]: I found the limitation is that compile cannot support DTensor constructor either in grad mode or no-grad mode, so we really need a separate path for compile :(

Hopefully we fix these tensor subclass papercuts very soon and can remove this divergence.

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_common.py:123]: Sounds good!
